File: lib/host/google.c
/* 
* Copyright (C) 2009,2010 Toni Gundogdu.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "host.h"

_host_constants(google, "video.google.", "flv|mp4");

_host_re(re_id,     "(?i)docid:'(.*?)'");
_host_re(re_title,  "(?i)<title>(.*?)</");
_host_re(re_lnk,    "(?i)videourl\\W+x3d(.*?)\\W+x26");
_host_re(re_mp4,    "(?i)href=(.*?)=ck1");

QUVIcode
handle_google(const char *url, _quvi_video_t video) {
    char *content, *lnk, *format;
    QUVIcode rc;

    /* host id */
    _host_id("google");

    /* common */
    rc = parse_page_common(url, video, &content, re_id, re_title);

    if (rc != QUVI_OK)
        return (rc);

    /* video link */
    lnk = 0;

    format = video->quvi->format;

    if (format) {
        if (!strcmp(format, "mp4") || !strcmp(format, "best")) {
            char *mp4;

            rc = regexp_capture(
                video->quvi,
                content,
                re_mp4,
                0,
                0,
                &mp4,
                (void *) 0
            );

            if (rc == QUVI_OK && mp4) {
                asprintf(&lnk, "%s=ck1", mp4);
                _free(mp4);
            }
        }
    }

    if (!lnk) { /* default to flv */
        rc = regexp_capture(
            video->quvi,
            content,
            re_lnk,
            0,
            0,
            &lnk,
            (void *) 0
        );

        if (rc != QUVI_OK)
            return (rc);
    }

    _free(content);

    lnk = unescape(video->quvi, lnk); /* orig. lnk freed by unescape */
    rc = add_video_link(&video->link, "%s", lnk);

    _free(lnk);

    return (rc);
}

File: share/lua/website/google.lua

-- quvi
-- Copyright (C) 2010  Toni Gundogdu <legatvs@gmail.com>
--
-- This file is part of quvi <http://quvi.sourceforge.net/>.
--
-- This library is free software; you can redistribute it and/or
-- modify it under the terms of the GNU Lesser General Public
-- License as published by the Free Software Foundation; either
-- version 2.1 of the License, or (at your option) any later version.
--
-- This library is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General Public
-- License along with this library; if not, write to the Free Software
-- Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
-- 02110-1301  USA
--

-- Identify the script.
function ident (self)
    package.path = self.script_dir .. '/?.lua'
    local C      = require 'quvi/const'
    local r      = {}
    r.domain     = "video.google."
    r.formats    = "default"
    r.categories = C.proto_http
    local U      = require 'quvi/util'
    r.handles    = U.handles(self.page_url,
                    {r.domain}, {"videoplay"}, {"docid=[%w-]+"})
    return r
end

-- Parse video URL.
function parse (self)
    self.host_id = "google"
    local page   = quvi.fetch(self.page_url)

    local _,_,s = page:find("<title>(.-)</title>")
    self.title  = s or error ("no match: video title")

    local _,_,s = page:find("docid:'(.-)'")
    self.id     = s or error ("no match: video id")

    local _,_,s = page:find("videoUrl%Wx3d(.-)%Wx26")
    s           = s or error ("no match: url")
    local U     = require 'quvi/util'
    self.url    = {U.unescape(s)}

    return self
end

-- vim: set ts=4 sw=4 tw=72 expandtab:
